homeassistant/components/vilfo/sensor.py

```diff
@@ -0,0 +1,91 @@
+"""Support for Vilfo Router sensors."""
+from homeassistant.helpers.entity import Entity
+
+from .const import (
+    ATTR_API_DATA_FIELD,
+    ATTR_DEVICE_CLASS,
+    ATTR_ICON,
+    ATTR_LABEL,
+    ATTR_UNIT,
+    DOMAIN,
+    ROUTER_DEFAULT_MODEL,
+    ROUTER_DEFAULT_NAME,
+    ROUTER_MANUFACTURER,
+    SENSOR_TYPES,
+)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Add Vilfo Router entities from a config_entry."""
+    vilfo = hass.data[DOMAIN][config_entry.entry_id]
+
+    sensors = []
+
+    device_info = {
+        "identifiers": {(DOMAIN, vilfo.host, vilfo.mac_address)},
+        "name": ROUTER_DEFAULT_NAME,
+        "manufacturer": ROUTER_MANUFACTURER,
+        "model": ROUTER_DEFAULT_MODEL,
+        "sw_version": vilfo.firmware_version,
+    }
+
+    for sensor_type in SENSOR_TYPES:
+        sensors.append(VilfoRouterSensor(sensor_type, vilfo, device_info))
+
+    async_add_entities(sensors, True)
+
+
+class VilfoRouterSensor(Entity):
+    """Define a Vilfo Router Sensor."""
+
+    def __init__(self, sensor_type, api, device_info):
+        """Initialize."""
+        self.api = api
+        self.sensor_type = sensor_type
+        self._device_info = device_info
```

Comment:
Initialize `device_info` here  or return directly in the property.

So you can remove the `device_info` param.

Reply:
Sorry, but I don't follow 😄 

**Do you mean refactor into something like:**

```python
# ...
def __init__(self, sensor_type, api, device_info):
    # ...
    self.device_info = device_info
# ...
```
(and then remove the `device_info` method)

**or do you mean:**

Pass all the attributes of the `device_info` object (defined in `async_setup_entry`) as individual parameters to the constructor (and then stitching them together as a single object again)?

Reply:
I mean : 

```suggestion
        self._device_info = {
        "identifiers": {(DOMAIN, vilfo.host, vilfo.mac_address)},
        "name": ROUTER_DEFAULT_NAME,
        "manufacturer": ROUTER_MANUFACTURER,
        "model": ROUTER_DEFAULT_MODEL,
        "sw_version": vilfo.firmware_version,
    }
```